// @slack/web-api@7.10.0/dist/types/response/ChatPostMessageResponse.js
"use strict";
/////////////////////////////////////////////////////////////////////////////////////////
//                                                                                     //
// !!! DO NOT EDIT THIS FILE !!!                                                       //
//                                                                                     //
// This file is auto-generated by scripts/generate-web-api-types.sh in the repository. //
// Please refer to the script code to learn how to update the source data.             //
//                                                                                     //
/////////////////////////////////////////////////////////////////////////////////////////
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockType = exports.AccessoryType = exports.PurpleType = exports.DescriptionType = void 0;
var DescriptionType;
(function (DescriptionType) {
    DescriptionType["Mrkdwn"] = "mrkdwn";
    DescriptionType["PlainText"] = "plain_text";
})(DescriptionType || (exports.DescriptionType = DescriptionType = {}));
var PurpleType;
(function (PurpleType) {
    PurpleType["Broadcast"] = "broadcast";
    PurpleType["Channel"] = "channel";
    PurpleType["Color"] = "color";
    PurpleType["Date"] = "date";
    PurpleType["Emoji"] = "emoji";
    PurpleType["Link"] = "link";
    PurpleType["Team"] = "team";
    PurpleType["Text"] = "text";
    PurpleType["User"] = "user";
    PurpleType["Usergroup"] = "usergroup";
})(PurpleType || (exports.PurpleType = PurpleType = {}));
var AccessoryType;
(function (AccessoryType) {
    AccessoryType["Button"] = "button";
    AccessoryType["ChannelsSelect"] = "channels_select";
    AccessoryType["Checkboxes"] = "checkboxes";
    AccessoryType["ConversationsSelect"] = "conversations_select";
    AccessoryType["Datepicker"] = "datepicker";
    AccessoryType["Datetimepicker"] = "datetimepicker";
    AccessoryType["ExternalSelect"] = "external_select";
    AccessoryType["Image"] = "image";
    AccessoryType["MultiChannelsSelect"] = "multi_channels_select";
    AccessoryType["MultiConversationsSelect"] = "multi_conversations_select";
    AccessoryType["MultiExternalSelect"] = "multi_external_select";
    AccessoryType["MultiStaticSelect"] = "multi_static_select";
    AccessoryType["MultiUsersSelect"] = "multi_users_select";
    AccessoryType["Overflow"] = "overflow";
    AccessoryType["RadioButtons"] = "radio_buttons";
    AccessoryType["RichTextList"] = "rich_text_list";
    AccessoryType["RichTextPreformatted"] = "rich_text_preformatted";
    AccessoryType["RichTextQuote"] = "rich_text_quote";
    AccessoryType["RichTextSection"] = "rich_text_section";
    AccessoryType["StaticSelect"] = "static_select";
    AccessoryType["Timepicker"] = "timepicker";
    AccessoryType["UsersSelect"] = "users_select";
    AccessoryType["WorkflowButton"] = "workflow_button";
})(AccessoryType || (exports.AccessoryType = AccessoryType = {}));
var BlockType;
(function (BlockType) {
    BlockType["Actions"] = "actions";
    BlockType["Context"] = "context";
    BlockType["Divider"] = "divider";
    BlockType["Image"] = "image";
    BlockType["RichText"] = "rich_text";
    BlockType["Section"] = "section";
    BlockType["ShareShortcut"] = "share_shortcut";
    BlockType["Video"] = "video";
})(BlockType || (exports.BlockType = BlockType = {}));
