// @slack/web-api@7.10.0/dist/types/response/FilesRemoteAddResponse.js
"use strict";
/////////////////////////////////////////////////////////////////////////////////////////
//                                                                                     //
// !!! DO NOT EDIT THIS FILE !!!                                                       //
//                                                                                     //
// This file is auto-generated by scripts/generate-web-api-types.sh in the repository. //
// Please refer to the script code to learn how to update the source data.             //
//                                                                                     //
/////////////////////////////////////////////////////////////////////////////////////////
Object.defineProperty(exports, "__esModule", { value: true });
exports.FluffyType = exports.PurpleType = exports.TextType = void 0;
var TextType;
(function (TextType) {
    TextType["Mrkdwn"] = "mrkdwn";
    TextType["PlainText"] = "plain_text";
})(TextType || (exports.TextType = TextType = {}));
var PurpleType;
(function (PurpleType) {
    PurpleType["Broadcast"] = "broadcast";
    PurpleType["Channel"] = "channel";
    PurpleType["Color"] = "color";
    PurpleType["Date"] = "date";
    PurpleType["Emoji"] = "emoji";
    PurpleType["Link"] = "link";
    PurpleType["Team"] = "team";
    PurpleType["Text"] = "text";
    PurpleType["User"] = "user";
    PurpleType["Usergroup"] = "usergroup";
})(PurpleType || (exports.PurpleType = PurpleType = {}));
var FluffyType;
(function (FluffyType) {
    FluffyType["RichTextList"] = "rich_text_list";
    FluffyType["RichTextPreformatted"] = "rich_text_preformatted";
    FluffyType["RichTextQuote"] = "rich_text_quote";
    FluffyType["RichTextSection"] = "rich_text_section";
})(FluffyType || (exports.FluffyType = FluffyType = {}));
