// @slack/web-api@7.10.0/dist/types/response/SearchFilesResponse.js
"use strict";
/////////////////////////////////////////////////////////////////////////////////////////
//                                                                                     //
// !!! DO NOT EDIT THIS FILE !!!                                                       //
//                                                                                     //
// This file is auto-generated by scripts/generate-web-api-types.sh in the repository. //
// Please refer to the script code to learn how to update the source data.             //
//                                                                                     //
/////////////////////////////////////////////////////////////////////////////////////////
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ts = exports.UserTeam = exports.LastEditor = exports.BlockType = exports.FluffyType = exports.PurpleType = exports.DescriptionType = void 0;
var DescriptionType;
(function (DescriptionType) {
    DescriptionType["Mrkdwn"] = "mrkdwn";
    DescriptionType["PlainText"] = "plain_text";
})(DescriptionType || (exports.DescriptionType = DescriptionType = {}));
var PurpleType;
(function (PurpleType) {
    PurpleType["Broadcast"] = "broadcast";
    PurpleType["Channel"] = "channel";
    PurpleType["Color"] = "color";
    PurpleType["Date"] = "date";
    PurpleType["Emoji"] = "emoji";
    PurpleType["Link"] = "link";
    PurpleType["Team"] = "team";
    PurpleType["Text"] = "text";
    PurpleType["User"] = "user";
    PurpleType["Usergroup"] = "usergroup";
})(PurpleType || (exports.PurpleType = PurpleType = {}));
var FluffyType;
(function (FluffyType) {
    FluffyType["RichTextList"] = "rich_text_list";
    FluffyType["RichTextPreformatted"] = "rich_text_preformatted";
    FluffyType["RichTextQuote"] = "rich_text_quote";
    FluffyType["RichTextSection"] = "rich_text_section";
})(FluffyType || (exports.FluffyType = FluffyType = {}));
var BlockType;
(function (BlockType) {
    BlockType["Actions"] = "actions";
    BlockType["Context"] = "context";
    BlockType["Divider"] = "divider";
    BlockType["Image"] = "image";
    BlockType["RichText"] = "rich_text";
    BlockType["Section"] = "section";
    BlockType["ShareShortcut"] = "share_shortcut";
    BlockType["Video"] = "video";
})(BlockType || (exports.BlockType = BlockType = {}));
var LastEditor;
(function (LastEditor) {
    LastEditor["Empty"] = "";
    LastEditor["U00000000"] = "U00000000";
})(LastEditor || (exports.LastEditor = LastEditor = {}));
var UserTeam;
(function (UserTeam) {
    UserTeam["Empty"] = "";
    UserTeam["T00000000"] = "T00000000";
})(UserTeam || (exports.UserTeam = UserTeam = {}));
var Ts;
(function (Ts) {
    Ts["Empty"] = "";
    Ts["The0000000000000000"] = "0000000000.000000";
})(Ts || (exports.Ts = Ts = {}));
