// @slack/web-api@7.10.0/dist/types/response/SearchMessagesResponse.js
"use strict";
/////////////////////////////////////////////////////////////////////////////////////////
//                                                                                     //
// !!! DO NOT EDIT THIS FILE !!!                                                       //
//                                                                                     //
// This file is auto-generated by scripts/generate-web-api-types.sh in the repository. //
// Please refer to the script code to learn how to update the source data.             //
//                                                                                     //
/////////////////////////////////////////////////////////////////////////////////////////
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockType = exports.PurpleType = exports.DescriptionType = exports.ActionType = void 0;
var ActionType;
(function (ActionType) {
    ActionType["Button"] = "button";
    ActionType["ChannelsSelect"] = "channels_select";
    ActionType["Checkboxes"] = "checkboxes";
    ActionType["ConversationsSelect"] = "conversations_select";
    ActionType["Datepicker"] = "datepicker";
    ActionType["Datetimepicker"] = "datetimepicker";
    ActionType["ExternalSelect"] = "external_select";
    ActionType["Image"] = "image";
    ActionType["MultiChannelsSelect"] = "multi_channels_select";
    ActionType["MultiConversationsSelect"] = "multi_conversations_select";
    ActionType["MultiExternalSelect"] = "multi_external_select";
    ActionType["MultiStaticSelect"] = "multi_static_select";
    ActionType["MultiUsersSelect"] = "multi_users_select";
    ActionType["Overflow"] = "overflow";
    ActionType["RadioButtons"] = "radio_buttons";
    ActionType["RichTextList"] = "rich_text_list";
    ActionType["RichTextPreformatted"] = "rich_text_preformatted";
    ActionType["RichTextQuote"] = "rich_text_quote";
    ActionType["RichTextSection"] = "rich_text_section";
    ActionType["StaticSelect"] = "static_select";
    ActionType["Timepicker"] = "timepicker";
    ActionType["UsersSelect"] = "users_select";
    ActionType["WorkflowButton"] = "workflow_button";
})(ActionType || (exports.ActionType = ActionType = {}));
var DescriptionType;
(function (DescriptionType) {
    DescriptionType["Mrkdwn"] = "mrkdwn";
    DescriptionType["PlainText"] = "plain_text";
})(DescriptionType || (exports.DescriptionType = DescriptionType = {}));
var PurpleType;
(function (PurpleType) {
    PurpleType["Broadcast"] = "broadcast";
    PurpleType["Channel"] = "channel";
    PurpleType["Color"] = "color";
    PurpleType["Date"] = "date";
    PurpleType["Emoji"] = "emoji";
    PurpleType["Link"] = "link";
    PurpleType["Team"] = "team";
    PurpleType["Text"] = "text";
    PurpleType["User"] = "user";
    PurpleType["Usergroup"] = "usergroup";
})(PurpleType || (exports.PurpleType = PurpleType = {}));
var BlockType;
(function (BlockType) {
    BlockType["Actions"] = "actions";
    BlockType["Context"] = "context";
    BlockType["Divider"] = "divider";
    BlockType["Image"] = "image";
    BlockType["RichText"] = "rich_text";
    BlockType["Section"] = "section";
    BlockType["ShareShortcut"] = "share_shortcut";
    BlockType["Video"] = "video";
})(BlockType || (exports.BlockType = BlockType = {}));
