// @slack/web-api/dist/response/SearchAllResponse.js
"use strict";
/* eslint-disable */
/////////////////////////////////////////////////////////////////////////////////////////
//                                                                                     //
// !!! DO NOT EDIT THIS FILE !!!                                                       //
//                                                                                     //
// This file is auto-generated by scripts/generate-web-api-types.sh in the repository. //
// Please refer to the script code to learn how to update the source data.             //
//                                                                                     //
/////////////////////////////////////////////////////////////////////////////////////////
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ts = exports.Team = exports.LastEditor = void 0;
var LastEditor;
(function (LastEditor) {
    LastEditor["Empty"] = "";
    LastEditor["U00000000"] = "U00000000";
})(LastEditor = exports.LastEditor || (exports.LastEditor = {}));
var Team;
(function (Team) {
    Team["Empty"] = "";
    Team["T00000000"] = "T00000000";
})(Team = exports.Team || (exports.Team = {}));
var Ts;
(function (Ts) {
    Ts["Empty"] = "";
    Ts["The0000000000000000"] = "0000000000.000000";
})(Ts = exports.Ts || (exports.Ts = {}));
